### PR TITLE
added download artifacts task for circleCI

### PR DESCRIPTION
**Pull request by PranaySPatil**

### CODE REVIEW

File: Tasks/CircleCIDownloadArtifacts/commitsDownloader.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    private GetCommitsFromSingleBuild(buildId: any): Q.Promise<string> {
+        let connection = tl.getInput("connection", true);
+        let definitionId = tl.getInput("definition", true);
+        var endpointUrl = tl.getEndpointUrl(connection, false);
+        let defer = Q.defer<string>();
+        let url: string = endpointUrl + "/api/v1.1/project/"+definitionId+"/"+buildId;
+        
+        this.webProvider.webClient.get(url, { 'Accept': 'application/json' }).then((res: HttpClientResponse) => {
```
Comment (omeshp) [inline Tasks/CircleCIDownloadArtifacts/commitsDownloader.ts]: add retires.

File: Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+        return defer.promise;
+    }
+
+    private GetCommitsFromSingleBuild(buildId: any): Q.Promise<string> {
+        let connection = tl.getInput("connection", true);
+        let definitionId = tl.getInput("definition", true);
+        var endpointUrl = tl.getEndpointUrl(connection, false);
+        let defer = Q.defer<string>();
+        let url: string = endpointUrl + "/api/v1.1/project/"+definitionId+"/"+buildId;
```
Comment (omeshp) [inline Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts]: handle double slashes?

File: Tasks/CircleCIDownloadArtifactsV0/download.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        let downloadPath = tl.getInput("downloadPath", true);
+        var retryLimit = parseInt(tl.getVariable("VSTS_HTTP_RETRY")) ? parseInt(tl.getVariable("VSTS_HTTP_RETRY")) : 4;
+
+        var endpointUrl = tl.getEndpointUrl(connection, false);
+        var itemsUrl = endpointUrl + "/api/v1.1/project/"+definitionId+"/"+buildId+"/artifacts";
+        itemsUrl = itemsUrl.replace(/([^:]\/)\/+/g, "$1");
+        console.log(tl.loc("DownloadArtifacts", itemsUrl));
+
+        var templatePath = path.join(__dirname, 'circleCI.handlebars.txt');
+        var username = tl.getEndpointAuthorizationParameter(connection, 'username', false);
```
Comment (omeshp) [inline Tasks/CircleCIDownloadArtifactsV0/download.ts]: username isn't treated as secret in UI, how do we handle not exposing these secrets to users with view endpoint permission?

File: Tasks/CircleCIDownloadArtifactsV0/download.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    var promise = new Promise<void>(async (resolve, reject) => {
+        let connection = tl.getInput("connection", true);
+        let definitionId = tl.getInput("definition", true);
+        let buildId = tl.getInput("version", true);
+        let itemPattern = tl.getInput("itemPattern", false);
+        let downloadPath = tl.getInput("downloadPath", true);
+        var retryLimit = parseInt(tl.getVariable("VSTS_HTTP_RETRY")) ? parseInt(tl.getVariable("VSTS_HTTP_RETRY")) : 4;
+
+        var endpointUrl = tl.getEndpointUrl(connection, false);
+        var itemsUrl = endpointUrl + "/api/v1.1/project/"+definitionId+"/"+buildId+"/artifacts";
```
Comment (omeshp) [inline Tasks/CircleCIDownloadArtifactsV0/download.ts]:
```suggestion
        var itemsUrl = endpointUrl + "/api/v1.1/project/" + definitionId + "/" + buildId + "/artifacts";
```

File: Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        
+        this.executeWithRetries("getCommitsFromSingleBuild", this.webProvider.webClient.get, 4, url, { 'Accept': 'application/json' }).then((res: HttpClientResponse) => {
+            if (res && res.message.statusCode === 200) {
```
Comment (omeshp) [inline Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts]: where is the reject clause?

File: Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    private GetCommits(startBuildId: string, endBuildId: string): Q.Promise<string> {
+        let connection = tl.getInput("connection", true);
+        let definitionId = tl.getInput("definition", true);
+        var endpointUrl = tl.getEndpointUrl(connection, false);
+        let defer = Q.defer<string>();
+        let url: string = endpointUrl + "/api/v1.1/project/"+definitionId+"/"+startBuildId;
+        let commits = [];
+
+        this.executeWithRetries("getCommits", this.webProvider.webClient.get, 4, url, { 'Accept': 'application/json' }).then((res: HttpClientResponse) => {
+            if (res && res.message.statusCode === 200) {
```
Comment (omeshp) [inline Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts]: no reject clause?

File: Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+                res.readBody().then((body: string) => {
+                    let jsonResult = JSON.parse(body);
+                    let branch = jsonResult.branch;
+
+                    let buildsUrl = endpointUrl + "/api/v1.1/project/"+definitionId+"/tree/"+branch;
```
Comment (omeshp) [inline Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts]:
```suggestion
                    let buildsUrl = endpointUrl + "/api/v1.1/project/" + definitionId + "/tree/" + branch;
```

File: Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts (hunk 2 above)
Comment (kasubram) [reply Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts]: can you use string formater?
`${endpointUrl}/api/v1.1/project/${definitionId}/${buildId};

Comment (kasubram) [reply Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts]: also chomp of trailing slash in endpoint which is most likely.

File: Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts (hunk 7 above)
Comment (kasubram) [reply Tasks/CircleCIDownloadArtifactsV0/commitsdownloader.ts]: please use a string formatter like mentioned above

File: Tasks/CircleCIDownloadArtifactsV0/download.ts (hunk 4 above)
Comment (kasubram) [reply Tasks/CircleCIDownloadArtifactsV0/download.ts]: formatter

File: Tasks/CircleCIDownloadArtifactsV0/download.ts (hunk 3 above)
Comment (PranaySPatil) [reply Tasks/CircleCIDownloadArtifactsV0/download.ts]: added isconfidential prop for username field in the endpoint

**Pull request merged**